base/util.go

```diff
@@ -1056,3 +1057,70 @@ func Sha1HashString(str string, salt string) string {
 	hashedKey := h.Sum(nil)
 	return fmt.Sprintf("%x", hashedKey)
 }
+
+// InjectJSONProperty takes the given JSON byte slice, marshals val, and inserts under the given key, and returns the result.
+func InjectJSONProperty(b []byte, key string, val interface{}) (new []byte, err error) {
+	bIsJSONObject, bIsEmpty := isJSONObject(b)
+	if !bIsJSONObject {
+		return nil, errors.New("b is not a JSON object")
+	}
+
+	valBytes, err := json.Marshal(val)
+	if err != nil {
+		return nil, err
+	}
+
+	return injectJSONPropertyFromBytes(b, bIsEmpty, key, valBytes), nil
+}
+
+// InjectJSONPropertyFromBytes takes the given JSON byte slice, and inserts val under the given key, and returns the result.
+func InjectJSONPropertyFromBytes(b []byte, key string, val []byte) (new []byte, err error) {
+	bIsJSONObject, bIsEmpty := isJSONObject(b)
+	if !bIsJSONObject {
+		return nil, errors.New("b is not a JSON object")
+	}
+
+	return injectJSONPropertyFromBytes(b, bIsEmpty, key, val), nil
+}
+
+// isJSONObject checks if the given bytes are a JSON object,
+// and also whether it's an empty object or not.
+func isJSONObject(b []byte) (isJSONObject, isEmpty bool) {
+
+	// Check if the byte slice starts with { and ends with }
+	if len(b) < 2 || b[0] != 0x7b || b[len(b)-1] != 0x7d {
+		return false, false
+	}
+
+	return true, len(b) == 2
+}
+
+// injectJSONPropertyFromBytes injects val under the given key into b, assuming b is a JSON object.
```

Comment:
As discussed, could do a quick review to evaluate whether there are any use cases where we'd want to be mutating the incoming byte slice, instead of allocating a new one.

Reply:
Current potential usages of this are to inject `_attachments` just before storing backup revision docs, and injecting `_sync` under non-xattr cases when writing to the bucket.

In both these cases, I don't think we want to modify the existing bytes, as they're stored on a shared Document during an update.

Reply:
Found a use case where mutating the original byte slice would be worth it... Injecting multiple properties into a document, without incurring a copy for each value. E.g:

```go
myDoc := []byte(`{"test":true}`)
if err := InjectJSONProperty(myDoc, "_id", docID); err != nil {
    return nil, err
}
if err := InjectJSONProperty(myDoc, "_rev", revID); err != nil {
    return nil, err
}
```